### PR TITLE
[pickers] Improve `PickerValidDate` type

### LINKED ISSUE #14765: DatePicker has incorrect type for value prop
### Search keywords

datepicker type value

### Latest version

- [X] I have tested the latest version

### Affected products

Date and Time Pickers

### Steps to reproduce

Link to live example: https://stackblitz.com/edit/github-gm5jav

It is very hard to create a shareable repro, but the component could be a simple `<DatePicker value={new Date()} />`

### Current behavior

`value` only accepts `null` or `undefined`

### Expected behavior

To have the correct type

### Context

It is very hard to create a shareable repro, but the issue is that we have a monorepo with a couple of react apps, and a separate package for our common stuff such as ui elements and forms, some of which have DatePickers.

However, because we're not surrounding our shared components with a LocalizationProvider, the type for `DatePicker.value` becomes `never`. (but the packages that use them do actually have LocalizationProviders that are setup correctly)

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: macOS 15.0
  Binaries:
    Node: 21.7.3 - ~/.nvm/versions/node/v21.7.3/bin/node
    npm: 10.5.0 - ~/.nvm/versions/node/v21.7.3/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 129.0.6668.70
    Edge: Not Found
    Safari: 18.0
```
</details>

### Order ID or Support key 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by flaviendelangle**
Fixes #14765

If no adapter has registered a date type in `PickerValidDate`, then we fallback to `any` instead of `never`.

### CODE REVIEW

**Comment by flaviendelangle** [comment]
@LukasTy how would you handle the doc section?
For people with the new version of the package, it make no sense since the error no longer exists.
But for people with "older" version they can still find it in the doc and it can help them.

Maybe we can let it as is and during the alpha we re-word it to remove the "error" aspect and just present how to correctly type if people have a "any" as the date type.

**Comment by LukasTy** [comment]
> Maybe we can let it as is and during the alpha we re-word it to remove the "error" aspect and just present how to correctly type if people have a "any" as the date type.

What do you think about updating the second paragraph of the doc section to reflect these new changes in this PR? 🤔 
Something like:
```
If you see `value` or other similar props typed as `any` it means that there is no `LocalizationProvider` with a defined adapter in your TypeScript project scope.
You can fix it by manually importing the adapter in some file of your project as follows:
```

**Comment by flaviendelangle** [comment]
> If you see `value` or other similar props typed as `any`

People won't directly see, they will see the type they pass to their prop.

To task a very dumb example:

image


But they can import `PickerValidDate` to check:

image


So maybe:


>If you are not sure your adapter is set up correctly to infer the type of the `value` prop (or other similar props), 
> you can import the `PickerValidDate` type and check its current value:
>
> image

**Comment by flaviendelangle** [comment]
The other solution I see is to remove the `TDate` generic and just use `PickerValidDate` everywhere.
In OUR codebase it would mean that we could no longer limit the valid value to be `Dayjs` or `Date` for example, because we have all the adapters loaded.
But in a real application it would simplify the typing and make your proposal possible.

**Comment by LukasTy** [comment]
> But in a real application it would simplify the typing and make your proposal possible.
> 
> This is typing breaking change, but one I'm totally willing to make 👍

That sounds great. Can we do this for v8? 🤔 
Shouldn't have too much of an impact for most users anyways. 👍

**Comment by flaviendelangle** [comment]
> Can we do this for v8? 🤔

Sure
I finish #14651 and we review it before the alpha
That way I can work on this PR during the alpha without massive Git conflicts on every single interface :laughing:

**Comment by flaviendelangle** [comment]
@LukasTy are you good merging this PR as is or do you think we should keep it for v8 and the typing changes ?

**Comment by LukasTy** [comment]
> are you good merging this PR as is or do you think we should keep it for v8 and the typing changes ?

WDYT about the documentation situation?
Maybe we should go with your suggestions for the v7 and update if we change the usage to directly use the `PickerValidDate`? 🤔

**Comment by flaviendelangle** [comment]
:+1: Right, I'll update the doc

**Comment by flaviendelangle** [comment]
I tried to add some explanation on the doc :+1:

File: docs/data/date-pickers/base-concepts/base-concepts.md [hunk 1]
```diff
@@ -185,15 +185,28 @@ To correctly type all the props that are date-related, the adapters override a g
...
+If it's equal to `any`, you can fix it by manually importing the adapter in some file of your project as show below:
+
+<img src="/static/x/date-pickers/picker-valid-date-not-configured.png" alt="PickerValidDate not correctly configured" />
 
 ```ts
 // Replace `AdapterDayjs` with the adapter you are using.
 import type {} from '@mui/x-date-pickers/AdapterDayjs';
 ```
 
+:::success
```
Comment (LukasTy) [inline docs/data/date-pickers/base-concepts/base-concepts.md:203]: Nitpick: I would say that this is more of an `info` callout 🙈

**Review by LukasTy — APPROVED** [review]
Superb improvement on doc! 👍 💯

**Pull request merged**